tools/zulint/custom_rules.py

```diff
@@ -0,0 +1,190 @@
+                else:
+                    raise Exception("Invalid strip rule")
+            if pattern.search(line_to_check):
+                if rule.get("exclude_pattern"):
+                    if re.search(rule['exclude_pattern'], line_to_check):
+                        continue
+                print_err(identifier, color, '{} at {} line {}:'.format(
+                    rule['description'], fn, i+1))
+                print_err(identifier, color, line)
+                ok = False
+        except Exception:
+            print("Exception with %s at %s line %s" % (rule['pattern'], fn, i+1))
+            traceback.print_exc()
+
+    if exclude_lines:
+        print('Please remove exclusions for file %s: %s' % (fn, exclude_lines))
+
+    return ok
+
+def check_file_for_long_lines(fn: str,
+                              max_length: int,
+                              line_tups: List[LineTup]) -> bool:
+    ok = True
+    for (i, line, line_newline_stripped, line_fully_stripped) in line_tups:
+        if isinstance(line, bytes):
+            line_length = len(line.decode("utf-8"))
+        else:
+            line_length = len(line)
+        if (line_length > max_length and
+            '# type' not in line and 'test' not in fn and 'example' not in fn and
+            # Don't throw errors for markdown format URLs
+            not re.search(r"^\[[ A-Za-z0-9_:,&()-]*\]: http.*", line) and
+            # Don't throw errors for URLs in code comments
+            not re.search(r"[#].*http.*", line) and
+            not re.search(r"`\{\{ api_url \}\}[^`]+`", line) and
+                "# ignorelongline" not in line and 'migrations' not in fn):
+            print("Line too long (%s) at %s line %s: %s" % (len(line), fn, i+1, line_newline_stripped))
+            ok = False
+    return ok
+
+def custom_check_file(fn: str,
+                      identifier: str,
+                      rules: RuleList,
+                      color: Optional[Iterable[str]],
+                      max_length: Optional[int]=None) -> bool:
+    failed = False
+
+    line_tups = get_line_info_from_file(fn=fn)
+
+    rules_to_apply = get_rules_applying_to_fn(fn=fn, rules=rules)
+
+    for rule in rules_to_apply:
+        ok = check_file_for_pattern(
+            fn=fn,
+            line_tups=line_tups,
+            identifier=identifier,
+            color=color,
+            rule=rule,
+        )
+        if not ok:
+            failed = True
+
+    # TODO: Move the below into more of a framework.
+    firstline = None
+    lastLine = None
+    if line_tups:
+        firstline = line_tups[0][3]  # line_fully_stripped for the first line.
+        lastLine = line_tups[-1][1]
+
+    if max_length is not None:
+        ok = check_file_for_long_lines(
+            fn=fn,
+            max_length=max_length,
+            line_tups=line_tups,
+        )
+        if not ok:
+            failed = True
+
+    if firstline:
+        shebang_rules = [
```

Comment:
To extract `shebang_rules` out of `custom_check_file`, we will have to add a new parameter to this function (if we want to want to keep the shebang_rules check only for the first line) and this is something we will have to do for `build_custom_checkers` too while extracting zulip specific config from them.

So, I think we should combine the functions for custom_check into a common class. Then it will be easier to customise the rules; since we can then initialise the class in `custom_check.py` and modify rules like `shebang_rules` easily since they would then be the properties of the class. This should be better in the long run, right?

Reply:
So, one thing that may be tricky about trying to do a class is the fact that we construct rule sets from other rule sets.  I'm not sure what the right structure is.  

It also seems like `markdown_docs_length_exclude` should just be a `length_exclude` option?

Here's a half-baked idea where we make `RuleList` into a class:

```
custom_checks = {                                                                                               
    'js': RuleList(                                                                                             
        {'pattern': 'subject|SUBJECT',                                                                          
         'exclude': set(['static/js/util.js',                                                                   
                         'frontend_tests/']),                                                                   
         'exclude_pattern': 'emails',                                                                           
         'description': 'avoid subject in JS code',                                                             
         'good_lines': ['topic_name'],                                                                          
         'bad_lines': ['subject="foo"', ' MAX_SUBJECT_LEN']},                                                   
        + whitespace_rules + comma_whitespace_rule,                                                             
    max_length=120,                                                                                             
    length_exclude=[],                                                                                          
    shebang_rules=shebang_rules])                                                                               
```

Reply:
I am thinking something else here, I will try it out and see.